internal/worker/apiaddresssetter/worker.go

```diff
@@ -0,0 +1,334 @@
+type ControllerNodeService interface {
+	// WatchControllerNodes returns a watcher that observes changes to the
+	// controller nodes.
+	WatchControllerNodes() (watcher.NotifyWatcher, error)
+
+	// GetControllerIDs returns the list of controller IDs from the controller node
+	// records.
+	GetControllerIDs(ctx context.Context) ([]string, error)
+
+	// SetAPIAddresses sets the provided addresses associated with the provided
+	// controller ID.
+	//
+	// The following errors can be expected:
+	// - [controllernodeerrors.NotFound] if the controller node does not exist.
+	SetAPIAddresses(ctx context.Context, controllerID string, addrs network.SpaceHostPorts, mgmtSpace network.SpaceInfo) error
+}
+
+// ApplicationService is an interface for the application domain service.
+type ApplicationService interface {
+	// WatchNetNodeAddress watches for changes to the specified net nodes
+	// addresses.
+	// This notifies on any changes to the net nodes addresses. It is up to the
+	// caller to determine if the addresses they're interested in has changed.
+	WatchNetNodeAddress(ctx context.Context, netNodeUUIDs ...string) (watcher.NotifyWatcher, error)
+
+	// GetUnitNetNodes returns the net node UUIDs associated with the specified
+	// unit. The net nodes are selected in the same way as in GetUnitAddresses, i.e.
+	// the union of the net nodes of the cloud service (if any) and the net node
+	// of the unit.
+	//
+	// The following errors may be returned:
+	// - [uniterrors.UnitNotFound] if the unit does not exist
+	GetUnitNetNodes(ctx context.Context, unitName unit.Name) ([]string, error)
+
+	// GetUnitPublicAddresses returns all public addresses for the specified unit.
+	//
+	// The following errors may be returned:
+	// - [uniterrors.UnitNotFound] if the unit does not exist
+	// - [network.NoAddressError] if the unit has no public address associated
+	GetUnitPublicAddresses(ctx context.Context, unitName unit.Name) (network.SpaceAddresses, error)
+}
+
+// NetworkService is the interface that is used to interact with the
+// network spaces/subnets.
+type NetworkService interface {
+	// SpaceByName returns a space from state that matches the input name. If the
+	// space is not found, an error is returned matching
+	// [github.com/juju/juju/domain/network/errors.SpaceNotFound].
+	SpaceByName(ctx context.Context, name string) (*network.SpaceInfo, error)
+}
+
+// apiAddressSetterWorker is a worker which sets the API addresses for the
+// controller, watching for changes both in the controller node's ip addresses
+// and the controller config (the juju-mgmt-space key) to filter the addresses
+// based on the management space.
+type apiAddressSetterWorker struct {
+	catacomb catacomb.Catacomb
+
+	config Config
+
+	controllerNodeChanges chan struct{}
+
+	// controllerTrackers holds the workers which track the nodes we
+	// are currently watching (all the controller nodes).
+	controllerTrackers map[string]*controllerTracker
```

Comment:
Hell no.

Use a runner.

Reply:
Yeah, I initially did it with a runner, but then I checked what was being done in the peergrouper (https://github.com/juju/juju/blob/main/internal/worker/peergrouper/worker.go#L522) and it misled me. 
Changed it back to a runner now.